src/transformers/modeling_mobilebert.py

```diff
@@ -0,0 +1,1547 @@
+        array = tf.train.load_variable(tf_path, name)
+        names.append(name)
+        arrays.append(array)
+
+    for name, array in zip(names, arrays):
+        name = name.replace("ffn_layer", "ffn")
+        name = name.replace("FakeLayerNorm", "LayerNorm")
+        name = name.replace("extra_output_weights", "dense/kernel")
+        name = name.replace("bert", "mobilebert")
+        name = name.split("/")
+        # adam_v and adam_m are variables used in AdamWeightDecayOptimizer to calculated m and v
+        # which are not required for using pretrained model
+        if any(
+            n in ["adam_v", "adam_m", "AdamWeightDecayOptimizer", "AdamWeightDecayOptimizer_1", "global_step"]
+            for n in name
+        ):
+            logger.info("Skipping {}".format("/".join(name)))
+            continue
+        pointer = model
+        for m_name in name:
+            if re.fullmatch(r"[A-Za-z]+_\d+", m_name):
+                scope_names = re.split(r"_(\d+)", m_name)
+            else:
+                scope_names = [m_name]
+            if scope_names[0] == "kernel" or scope_names[0] == "gamma":
+                pointer = getattr(pointer, "weight")
+            elif scope_names[0] == "output_bias" or scope_names[0] == "beta":
+                pointer = getattr(pointer, "bias")
+            elif scope_names[0] == "output_weights":
+                pointer = getattr(pointer, "weight")
+            elif scope_names[0] == "squad":
+                pointer = getattr(pointer, "classifier")
+            else:
+                try:
+                    pointer = getattr(pointer, scope_names[0])
+                except AttributeError:
+                    logger.info("Skipping {}".format("/".join(name)))
+                    continue
+            if len(scope_names) >= 2:
+                num = int(scope_names[1])
+                pointer = pointer[num]
+        if m_name[-11:] == "_embeddings":
+            pointer = getattr(pointer, "weight")
+        elif m_name == "kernel":
+            array = np.transpose(array)
+        try:
+            assert pointer.shape == array.shape
+        except AssertionError as e:
+            e.args += (pointer.shape, array.shape)
+            raise
+        logger.info("Initialize PyTorch weight {}".format(name))
+        pointer.data = torch.from_numpy(array)
+    return model
+
+
+def mish(x):
+    return x * torch.tanh(nn.functional.softplus(x))
+
+
+class ManualLayerNorm(nn.Module):
+    def __init__(self, feat_size, eps=1e-6):
+        super().__init__()
+        self.bias = nn.Parameter(torch.zeros(feat_size))
+        self.weight = nn.Parameter(torch.ones(feat_size))
+        self.eps = eps
+
+    def forward(self, input_tensor):
+        mean = input_tensor.mean(-1, keepdim=True)
+        std = input_tensor.std(-1, keepdim=True)
+        return self.weight * (input_tensor - mean) / (std + self.eps) + self.bias
+
+
+class NoNorm(nn.Module):
```

Comment:
Is this just a linear layer?

Reply:
Yes, quoting the paper:

> By model latency analysis, we find that layer normalization (Ba et al., 2016) and gelu activation (Hendrycks and Gimpel, 2016) accounted for a considerable proportion of total latency. Therefore, we propose to replace them with new operations in our MobileBERT.
> 
> **Remove layer normalization** 
> 
> We replace the layer normalization of a n-channel hidden state h with an element wise linear transformation: NoNorm(h) = γ ◦ h + β, (1) where γ, β ∈ R n and ◦ denotes the Hadamard product. Please note that NoNorm has different properties from LayerNorm even in test mode since the original layer normalization is not a linear operation for a batch of vectors. 

Here's a table showing the results latency-wise:

![Screen Capture_select-area_20200615194228](https://user-images.githubusercontent.com/30755778/84716165-61b15200-af40-11ea-826a-6015e976fb0c.png)

https://arxiv.org/pdf/2004.02984.pdf

Reply:
This is basically elementwise multiplication and addition. Granted, in the actual implementation the parameters are 1D with the dimension corresponding to hidden size (H), and they get broadcasted to (CxH) during runtime where C is the input sequence length, but still this operation cannot be expressed as either Linear or Conv1d without looking hacky and confusing. AFAIK such an operation is also not present among PyTorch standard modules - would make life easier if it were.